AdyenCardScanner/Sources/CardImageParser.swift

```diff
@@ -0,0 +1,233 @@
+
+    // MARK: - Initializers
+
+    init() {
+        self.dateFormatter = DateFormatter()
+        self.dateFormatter.locale = Locale.current
+        self.dateFormatter.timeZone = TimeZone(secondsFromGMT: 0)
+    }
+
+    // MARK: - CardImageParsing
+
+    func parse(
+        image: CIImage,
+        completion: @escaping (CreditCard) -> Void
+    ) {
+        guard let transformedImage = transform(image: image) else { return }
+
+        let recognizeTextRequest = VNRecognizeTextRequest()
+        recognizeTextRequest.recognitionLevel = .accurate
+        recognizeTextRequest.usesLanguageCorrection = false
+
+        let imageRequestHandler = VNImageRequestHandler(ciImage: transformedImage, options: [:])
+        try? imageRequestHandler.perform([recognizeTextRequest])
+
+        guard let results = recognizeTextRequest.results, !results.isEmpty else {
+            return
+        }
+
+        let dispatchGroup = DispatchGroup()
+
+        dispatchGroup.enter()
+        DispatchQueue.global().async {
+            if self.cardNumber == nil {
+                self.cardNumber = self.extractCardNumber(from: results)
+            }
+            dispatchGroup.leave()
+        }
+
+        dispatchGroup.enter()
+        DispatchQueue.global().async {
+            if self.expireDate == nil {
+                self.expireDate = self.extractExpireDate(from: results)
+            }
+            dispatchGroup.leave()
+        }
+
+        dispatchGroup.notify(queue: .main) {
+            guard let cardNumber = self.cardNumber, let expireDate = self.expireDate else { return }
+            let card = CreditCard(number: cardNumber, expireDate: expireDate)
+            completion(card)
+        }
+    }
+
+    // MARK: - Private
+
+    private func transform(image: CIImage) -> CIImage? {
+        var outputImage: CIImage? = image
+
+        // Apply noise reduction filter.
+        let noiseReductionFilter = CIFilter(name: "CINoiseReduction")
+        noiseReductionFilter?.setValue(outputImage, forKey: kCIInputImageKey)
+        noiseReductionFilter?.setValue(0.02, forKey: "inputNoiseLevel")
+        noiseReductionFilter?.setValue(0.4, forKey: "inputSharpness")
+        outputImage = noiseReductionFilter?.outputImage
+
+        // Apply CIColorControls filter.
+        let colorControlsFilter = CIFilter(name: "CIColorControls")
+        colorControlsFilter?.setValue(outputImage, forKey: kCIInputImageKey)
+        colorControlsFilter?.setValue(0.2, forKey: kCIInputBrightnessKey)
+        colorControlsFilter?.setValue(1.5, forKey: kCIInputContrastKey)
+        colorControlsFilter?.setValue(1.2, forKey: kCIInputSaturationKey)
+        outputImage = colorControlsFilter?.outputImage
+
+        // Apply sharpness enhancement filter.
+        let sharpenFilter = CIFilter(name: "CISharpenLuminance")
+        sharpenFilter?.setValue(outputImage, forKey: kCIInputImageKey)
+        sharpenFilter?.setValue(0.5, forKey: kCIInputSharpnessKey)
```

Comment:
BTW, use use `kCIInputSharpnessKey` here and on line "96":
I assume only "CINoiseReduction" might be enough 🤔